File: backend/python/app/utils/request_context.py
"""Process-wide request id context for distributed tracing.

A ``ContextVar`` holds the root request id for the work in flight; the logging
layer reads it at emit time so call sites never pass the id. The per-service
suffix is a process-global set once per service via :func:`set_service_suffix`.
"""

import contextvars
import re
import uuid
from dataclasses import dataclass
from typing import Any, Dict, Optional

HEADER_REQUEST_ID = "x-request-id"
ENVELOPE_REQUEST_ID = "requestId"
NO_CONTEXT = "-"

# `<objectId:24>-<nanoid:21>`
_MAX_ROOT_ID_LEN = 64
_UNSAFE_ID_CHARS = re.compile(r"[^A-Za-z0-9._:-]")


def sanitize_root_id(raw: Optional[str]) -> Optional[str]:
    if not raw:
        return None
    cleaned = _UNSAFE_ID_CHARS.sub("", raw)[:_MAX_ROOT_ID_LEN]
    return cleaned or None


@dataclass
class RequestContext:
    root_id: str


_ctx: contextvars.ContextVar[Optional[RequestContext]] = contextvars.ContextVar(
    "request_ctx", default=None
)

_service_suffix: str = ""


def set_service_suffix(suffix: str) -> None:
    global _service_suffix
    _service_suffix = suffix or ""


def get_service_suffix() -> str:
    return _service_suffix


CtxToken = contextvars.Token["Optional[RequestContext]"]


def set_context(root_id: str) -> CtxToken:
    return _ctx.set(RequestContext(root_id=root_id))


def get_context() -> Optional[RequestContext]:
    return _ctx.get()


def reset_context(token: Optional[CtxToken]) -> None:
    if token is None:
        return
    try:
        _ctx.reset(token)
    except (ValueError, LookupError):
        # Token created in a different context (e.g. across a thread boundary).
        _ctx.set(None)


def new_system_root() -> str:
    """Root id for a system/automation-initiated unit of work.

    The ``sys-`` prefix marks the work as machine-originated (scheduled sync,
    token refresh, migration, broker message with no inbound id), the way a
    user request carries its ``<userId>-`` prefix.
    """
    return f"sys-{uuid.uuid4().hex}"


def new_anon_root() -> str:
    """Fallback root id when an inbound request carries none."""
    return f"anon-{uuid.uuid4().hex}"


def current_display_id() -> str:
    """The id stamped into log lines: ``<root><service-suffix>``, or ``-`` if none."""
    ctx = _ctx.get()
    if ctx is None:
        return NO_CONTEXT
    return f"{ctx.root_id}{_service_suffix}"


def inject_request_headers(
    headers: Optional[Dict[str, str]] = None,
) -> Dict[str, str]:
    out: Dict[str, str] = dict(headers or {})
    ctx = _ctx.get()
    if ctx is not None:
        out.setdefault(HEADER_REQUEST_ID, ctx.root_id)
    return out


def inject_envelope(message: Dict[str, Any]) -> Dict[str, Any]:
    """Return a copy of ``message`` with the root id stamped in (no mutation)."""
    ctx = _ctx.get()
    if ctx is None:
        return message
    out = dict(message)
    out.setdefault(ENVELOPE_REQUEST_ID, ctx.root_id)
    return out


def context_from_envelope(message: Dict[str, Any]) -> RequestContext:
    # A message with no id is a propagation gap across the broker boundary
    # (the broker analogue of an inbound HTTP request with no header), not
    # self-initiated machine work — so it gets an `anon-` root, not `sys-`.
    root_id = sanitize_root_id(message.get(ENVELOPE_REQUEST_ID))
    return RequestContext(root_id=root_id or new_anon_root())

File: backend/nodejs/apps/src/libs/context/request-context.ts
import { AsyncLocalStorage } from 'async_hooks';
import { randomUUID } from 'crypto';

/**
 * Ambient request-id context (AsyncLocalStorage). Mirrors the Python side
 * (`app/utils/request_context.py`): same header name and envelope key.
 */

export interface RequestContextStore {
  rootId: string;
}

export const HEADER_REQUEST_ID = 'x-request-id';
export const ENVELOPE_REQUEST_ID = 'requestId';
export const NO_CONTEXT = '-';

// `<objectId:24>-<nanoid:21>`
const MAX_ROOT_ID_LEN = 64;
const UNSAFE_ID_CHARS = /[^A-Za-z0-9._:-]/g;

/** Strip a client-supplied id to a safe charset + length. `undefined` if empty. */
export function sanitizeRootId(raw: string | undefined | null): string | undefined {
  if (!raw) return undefined;
  const cleaned = raw.replace(UNSAFE_ID_CHARS, '').slice(0, MAX_ROOT_ID_LEN);
  return cleaned || undefined;
}

const storage = new AsyncLocalStorage<RequestContextStore>();

/** Run `fn` with the given context bound for its entire async subtree. */
export function runWithRequestContext<T>(
  ctx: RequestContextStore,
  fn: () => T,
): T {
  return storage.run(ctx, fn);
}

export function getRequestContext(): RequestContextStore | undefined {
  return storage.getStore();
}

export function getRootId(): string | undefined {
  return storage.getStore()?.rootId;
}

/**
 * The `sys-` prefix marks it as machine-originated work.
 */
export function newSystemRoot(): string {
  return `sys-${randomUUID().replace(/-/g, '')}`;
}

/** Fallback root id when an inbound request carries none (mirrors Python `new_anon_root`). */
export function newAnonRoot(): string {
  return `anon-${randomUUID().replace(/-/g, '')}`;
}

/** The id stamped into log lines: `<root><service-suffix>` (node suffix is empty). */
export function currentDisplayId(suffix = ''): string {
  const ctx = storage.getStore();
  if (!ctx) return NO_CONTEXT;
  return `${ctx.rootId}${suffix}`;
}

/** Add the root id to outbound HTTP headers (returns a copy, never mutates). */
export function injectRequestHeaders(
  headers: Record<string, string> = {},
): Record<string, string> {
  const out = { ...headers };
  const ctx = storage.getStore();
  if (ctx && !hasHeader(out, HEADER_REQUEST_ID)) {
    out[HEADER_REQUEST_ID] = ctx.rootId;
  }
  return out;
}

/** Stamp the root id onto an outbound broker envelope (returns a copy). */
export function injectEnvelope<T extends Record<string, unknown>>(message: T): T {
  const ctx = storage.getStore();
  if (ctx && message[ENVELOPE_REQUEST_ID] === undefined) {
    return { ...message, [ENVELOPE_REQUEST_ID]: ctx.rootId };
  }
  return message;
}

function hasHeader(headers: Record<string, string>, name: string): boolean {
  const lower = name.toLowerCase();
  return Object.keys(headers).some((k) => k.toLowerCase() === lower);
}
